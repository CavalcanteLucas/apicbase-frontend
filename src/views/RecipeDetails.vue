<template>
  <div id="recipe-detail">
    <div class="container">
      <div class="col-12 mx-auto">
        <div class="row" style="padding-top: 2rem;">
          <h2>
            <strong>{{ recipe.name }}</strong>
          </h2>
        </div>
        <div class="row" style="padding-top: 2rem;">
          <div class="col-10 mx-auto">
            <table
              class="table table-striped"
              aria-label="Recipes details table"
            >
              <thead>
                <th scope="col">Ingredient</th>
                <th scope="col">Amount</th>
                <th scope="col">Unit</th>
                <th scope="col">Cost</th>
                <th scope="col">U. cost</th>
              </thead>
              <tbody>
                <tr v-for="item in recipeDetails" :key="item.id">
                  <td class="align-middle ">{{ item.ingredient }}</td>
                  <td class="align-middle">{{ item.amount_per_recipe }}</td>
                  <td class="align-middle">{{ item.unit }}</td>
                  <td class="align-middle">&euro;{{ item.cost }}</td>
                  <td class="align-middle">&euro;{{ item.cost_per_amount }}</td>
                  <td class="align-middle">
                    <button
                      type="button"
                      id="deleteBtn"
                      class="btn btn-sm btn-danger"
                      @click="deleteRecipeItem(item)"
                    >
                      <img src="../assets/img/trash.svg" alt="Delete" />
                    </button>
                    <confirm-dialogue ref="confirmDialogue"></confirm-dialogue>
                  </td>
                </tr>
              </tbody>
            </table>
            <div class="col-12" style="padding-top: 2rem;">
              <form id="recipesForm" @submit.prevent="createRecipeFormula">
                <div class="row g-1">
                  <div class="col-md-2  ms-auto">
                    <select
                      id="selectedIngredient"
                      class="form-select"
                      aria-label="Select ingredient"
                      v-model="selectedIngredient.id"
                      required="required"
                    >
                      <option
                        v-for="ingredient in ingredients"
                        :value="ingredient.id"
                        :key="ingredient.id"
                        >{{ ingredient.name }} ({{ ingredient.unit }})</option
                      >
                    </select>
                  </div>
                  <div class="col-md-1 ">
                    <input
                      type="number"
                      class="form-control"
                      id="inputAmount"
                      placeholder="Amount"
                      aria-label="Amount"
                      v-model="selectedIngredient.amount"
                      min=".01"
                      step=".01"
                      required="required"
                    />
                  </div>
                  <div class="col-md-3 ">
                    <button
                      type="button"
                      id="clearBtn"
                      v-on:click="clearRecipeForm()"
                      class="btn btn-outline-secondary float-start"
                    >
                      <img src="../assets/img/backspace.svg" alt="Clear" />
                    </button>
                    <button
                      type="submit"
                      id="submitBtn"
                      class="btn btn-success float-start"
                      style="margin-left: 0.25rem;"
                    >
                      <img src="../assets/img/plus-circle.svg" alt="Add" />
                    </button>
                  </div>
                </div>
              </form>
            </div>
            <h5>
              <strong
                >Total cost: &euro;{{ recipeCost }}</strong
              >
            </h5>
          </div>
        </div>
      </div>
    </div>
  </div>
</template>

<script>
import ConfirmDialogue from "../components/ConfirmDialogue.vue";

export default {
  name: "RecipeDetails",
  components: { ConfirmDialogue },
  data() {
    return {
      recipe: {},
      recipeDetails: [],
      recipeCost: null,

      ingredients: [],
      selectedIngredient: {
        id: null,
        amount: null,
      },
    };
  },
  async created() {
    await this.getRecipe();
    await this.getRecipeDetails();
    await this.getRecipeCost();
    await this.getIngredients();
  },
  methods: {
    async getIngredients() {
      let response = await fetch(`${process.env.VUE_APP_BACKEND_API}/api/ingredients/`);
      this.ingredients = await response.json();
    },
    async getRecipe() {
      let response = await fetch(
        `${process.env.VUE_APP_BACKEND_API}/api/recipes/${this.$route.params.id}/`
      );
      this.recipe = await response.json();
    },
    async getRecipeDetails() {
      let response = await fetch(
        `${process.env.VUE_APP_BACKEND_API}/api/recipes/${this.$route.params.id}/details/`
      );
      this.recipeDetails = await response.json();
    },
    async getRecipeCost() {
      let response = await fetch(
        `${process.env.VUE_APP_BACKEND_API}/api/recipes/${this.$route.params.id}/cost/`
      );
      this.recipeCost = await response.json();
    },
    async deleteRecipeItem(item) {
      await this.getRecipe();
      await this.getRecipeDetails();
      await this.getRecipeCost();
      const ok = await this.$refs.confirmDialogue[0].show();
      if (ok) {
        await fetch(`${process.env.VUE_APP_BACKEND_API}/api/recipes-formulas/${item.id}/`, {
          method: "delete",
          headers: {
            "Content-Type": "application/json",
          },
          body: JSON.stringify(this.ingredient),
        });
      }
      await this.getRecipe();
      await this.getRecipeDetails();
      await this.getRecipeCost();
    },
    async createRecipeFormula() {
      await this.getRecipe();
      await this.getRecipeDetails();
      await this.getRecipeCost();
      await fetch(`${process.env.VUE_APP_BACKEND_API}/api/recipes-formulas/`, {
        method: "post",
        headers: {
          "Content-Type": "application/json",
        },
        body: JSON.stringify({
          recipe: this.$route.params.id,
          ingredient: this.selectedIngredient.id,
          amount_per_recipe: this.selectedIngredient.amount,
        }),
      });
      await this.getRecipe();
      await this.getRecipeDetails();
      await this.getRecipeCost();
    },
    clearRecipeForm() {
      this.selectedIngredient = {
        id: null,
        amount: null,
      };
    },
  },
};
</script>

<style>
h5 {
  text-align: right;
  padding-top: 3rem;
}
</style>
