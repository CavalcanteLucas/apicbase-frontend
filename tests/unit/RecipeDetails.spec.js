import Vue from "vue";
import { mount, config } from "@vue/test-utils";
import RecipeDetails from "../../src/views/RecipeDetails.vue";

config.showDeprecationWarnings = false;

describe("recipe details view", () => {
  let wrapper;
  beforeEach(async () => {
    wrapper = mount(RecipeDetails, {
      methods: {
        getRecipe: () => {
          // do nothing
        },
        getRecipeDetails: () => {
          // do nothing
        },
        getRecipeCost: () => {
          // do nothing
        },
        getIngredients: () => {
          // do nothing
        },
      },
    });
    await wrapper.setData({
      recipe: {
        name: "carrot cake",
      },
      ingredients: [
        {
          id: 0,
          article_number: "42",
          name: "carrot",
          cost_per_amount: "1.00",
          amount: "500",
          unit: "gram",
        },
        {
          id: 1,
          article_number: "43",
          name: "cream",
          cost_per_amount: "5.00",
          amount: "10",
          unit: "centiliter",
        },
      ],
    });
  });

  it("renders", () => {
    expect(wrapper.exists()).toBe(true);
    expect(wrapper).toMatchSnapshot();
  });

  it("has data", () => {
    expect(typeof RecipeDetails.data).toBe("function");
  });

  it("is a Vue instance", () => {
    expect(wrapper.isVueInstance()).toBeTruthy();
  });

  it("renders the correct title, table headers, and total cost", () => {
    expect(wrapper.html()).toContain("carrot cake");
    expect(wrapper.html()).toContain("Ingredient");
    expect(wrapper.html()).toContain("Amount");
    expect(wrapper.html()).toContain("Unit");
    expect(wrapper.html()).toContain("U. cost");
    expect(wrapper.html()).toContain("Total cost:");
  });

  it("has two input fields", () => {
    expect(wrapper.contains('input[id="inputAmount"]')).toBe(true);
    expect(wrapper.contains('select[id="selectedIngredient"]')).toBe(true);
  });

  it("has a clear input button, and a submit button", () => {
    expect(wrapper.contains('button[id="clearBtn"]')).toBe(true);
    expect(wrapper.contains('button[id="submitBtn"]')).toBe(true);
  });

  it("input fields load data", () => {
    expect(wrapper.vm.$data.selectedIngredient).toEqual({
      id: null,
      amount: null,
    });
    wrapper.find('select[id="selectedIngredient"]').setValue(0);
    wrapper.find('input[id="inputAmount"]').setValue(100);
    expect(wrapper.find('select[id="selectedIngredient"]').element.value).toBe(
      "0"
    );
    expect(wrapper.find('input[id="inputAmount"]').element.value).toBe("100");
    expect(wrapper.vm.$data.selectedIngredient).toEqual({
      id: 0,
      amount: "100",
    });
  });

  it("clear button wipes loaded data out", async () => {
    await wrapper.setData({
      selectedIngredient: {
        id: 0,
        amount: "100",
      },
    });

    expect(wrapper.vm.$data.selectedIngredient).toEqual({
      id: 0,
      amount: "100",
    });

    let clearBtn = wrapper.find('button[id="clearBtn"]');
    expect(clearBtn.exists()).toBe(true);
    expect(clearBtn.is("button")).toBe(true);
    await clearBtn.trigger("click");

    expect(wrapper.vm.$data.selectedIngredient).toEqual({
      id: null,
      amount: null,
    });

    expect(wrapper.find('select[id="selectedIngredient"]').element.value).toBe(
      ""
    );
    expect(wrapper.find('input[id="inputAmount"]').element.value).toBe("");
  });

  it("opens confirmation modal when delete button is pressed", async () => {
    expect(wrapper.findAll("tbody > tr").wrappers).toHaveLength(0);
    await wrapper.setData({
      recipeDetails: [
        {
          ingredient: "carrot",
          amount_per_recipe: "100",
          unit: "gram",
          cost: "0.2",
          cost_per_amount: "0.002",
        },
        {
          ingredient: "cream",
          amount_per_recipe: "50",
          unit: "centiliter",
          cost: "2",
          cost_per_amount: "0.04",
        },
      ],
    });

    expect(wrapper.findAll("tbody > tr").wrappers).toHaveLength(2);
    let deleteBtn = wrapper.find('button[id="deleteBtn"]');
    expect(deleteBtn.exists()).toBe(true);
    expect(deleteBtn.is("button")).toBe(true);
    let modal = wrapper.find('div[class="popup-modal"]');
    expect(modal.exists()).toBe(false);

    await deleteBtn.trigger("click");
    await Vue.nextTick();

    modal = wrapper.find('div[class="popup-modal"]');

    expect(modal.exists()).toBe(true);
    expect(modal.is("div")).toBe(true);
    expect(modal.text()).toContain("Are you sure about that?");
    expect(modal.text()).toContain("Nevermind..");
    expect(modal.text()).toContain("Just do it!");
  });

  it("closes confirmation modal when cancel button is pressed", async () => {
    await wrapper.setData({
      recipeDetails: [
        {
          ingredient: "carrot",
          amount_per_recipe: "100",
          unit: "gram",
          cost: "0.2",
          cost_per_amount: "0.002",
        },
        {
          ingredient: "cream",
          amount_per_recipe: "50",
          unit: "centiliter",
          cost: "2",
          cost_per_amount: "0.04",
        },
      ],
    });
    let deleteBtn = wrapper.find('button[id="deleteBtn"]');
    expect(deleteBtn.exists()).toBe(true);
    expect(deleteBtn.is("button")).toBe(true);
    await deleteBtn.trigger("click");
    await Vue.nextTick();

    let cancelBtn = wrapper.find('button[id="cancelBtn"]');
    expect(cancelBtn.exists()).toBe(true);
    expect(cancelBtn.is("button")).toBe(true);
    await cancelBtn.trigger("click");

    let modal = wrapper.find('div[class="popup-modal"]');
    expect(modal.exists()).toBe(false);
    cancelBtn = wrapper.find('button[id="cancelBtn"]');
    expect(cancelBtn.exists()).toBe(false);
  });
});
