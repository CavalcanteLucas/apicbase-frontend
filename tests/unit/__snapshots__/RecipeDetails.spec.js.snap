// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`recipe details view renders 1`] = `
<div id="recipe-detail">
  <div class="container">
    <div class="col-12 mx-auto">
      <div class="row" style="padding-top: 2rem;">
        <h2><strong>carrot cake</strong></h2>
      </div>
      <div class="row" style="padding-top: 2rem;">
        <div class="col-10 mx-auto">
          <table aria-label="Recipes details table" class="table table-striped">
            <thead>
              <th scope="col">Ingredient</th>
              <th scope="col">Amount</th>
              <th scope="col">Unit</th>
              <th scope="col">Cost</th>
              <th scope="col">U. cost</th>
            </thead>
            <tbody></tbody>
          </table>
          <div class="col-12" style="padding-top: 2rem;">
            <form id="recipesForm">
              <div class="row g-1">
                <div class="col-md-2  ms-auto"><select id="selectedIngredient" aria-label="Select ingredient" required="required" class="form-select">
                    <option value="0">carrot (gram)</option>
                    <option value="1">cream (centiliter)</option>
                  </select></div>
                <div class="col-md-1 "><input type="number" id="inputAmount" placeholder="Amount" aria-label="Amount" min=".01" step=".01" required="required" class="form-control"></div>
                <div class="col-md-3 "><button type="button" id="clearBtn" class="btn btn-outline-secondary float-start"><img src="../assets/img/backspace.svg" alt="Clear"></button> <button type="submit" id="submitBtn" class="btn btn-success float-start" style="margin-left: 0.25rem;"><img src="../assets/img/plus-circle.svg" alt="Add"></button></div>
              </div>
            </form>
          </div>
          <h5><strong>Total cost: â‚¬</strong></h5>
        </div>
      </div>
    </div>
  </div>
</div>
`;
